src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,6 +741,117 @@ private void ReturnReadBufferIfEmpty()
                 _decryptedBytesCount = 0;
                 _decryptedBytesOffset = 0;
             }
+            else if (_decryptedBytesCount == 0)
+            {
+                _decryptedBytesOffset = 0;
+            }
+        }
+
+        private bool haveFullTlsFrame(ref int frameSize)
+        {
+            if (_internalBufferCount < SecureChannel.ReadHeaderSize)
+            {
+                frameSize = int.MaxValue;
+                return false;
+            }
+
+            frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+            return _internalBufferCount >= frameSize;
+        }
+
+        private ValueTask<int> getFullFrameIfNeed<TIOAdapter>(TIOAdapter adapter)
+            where TIOAdapter : IReadWriteAdapter
+        {
+            int frameSize;
+            if (_internalBufferCount >= SecureChannel.ReadHeaderSize)
+            {
+                frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+                if (_internalBufferCount >= frameSize)
+                {
+                    return new ValueTask<int>(frameSize);
+                }
+            }
+            else
+            {
+                frameSize = int.MaxValue;
+            }
+
+            // We may have enough space to complete frame, but we may still do extra IO if the frame is small.
+            // So we will attempt larger read - that is trade of with extra copy.
+            // This may be updated at some point based on size of existing chunk, rented buffer and size of 'buffer'.
+            ResetReadBuffer();
+
+            // _internalOffset is 0 after ResetReadBuffer and we use _internalBufferCount to determined where to read.
+            while (_internalBufferCount < frameSize)
+            {
+                // We don't have enough bytes buffered, so issue an initial read to try to get enough.  This is
+                // done in this method both to better consolidate error handling logic (the first read is the special
+                // case that needs to differentiate reading 0 from > 0, and everything else needs to throw if it
+                // doesn't read enough), and to minimize the chances that in the common case the FillBufferAsync
+                // helper needs to yield and allocate a state machine.
+
+                ValueTask<int> t = adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount));
+                if (!t.IsCompletedSuccessfully)
+                {
+                    return InternalGetFullFrameIfNeed(adapter, t, frameSize);
```

Comment:
Why is this split out, as opposed to just awaiting the ReadAsync?

Reply:
The goal is to avoid allocating async state if the ValueTask Read can finish synchronously. The `getFullFrameIfNeed` is not marked as `async`  I don't know if this has measurable impact but this is similar to FillHandshakeBufferAsync.

Reply:
The async state isn't allocated unless the method actually goes async. So we aren't really saving anything here relative to just awaiting.

Reply:
Can we clean this up? I don't understand why we are doing this.

Reply:
Is the `FillHandshakeBufferAsync`wrong  as well? Perhaps @stephentoub has better explanation than me. I feel that if we don't like this pattern we should remove it from SslStream in other places as well.

Reply:
It depends how hot the path is.  There is a small but measurable microbenchmark difference between something that just does `return new ValueTask<int>(someValue)` and `return someValue` in an async ValueTask-returning method; if memory serves, the latter has around 2x the overhead... but we're talking nanoseconds, e.g. maybe 5ns for the former and 10ns for the latter, or something like that.  If these methods are so much more expensive than that such that differences of that scale won't matter, we should do the simpler / more maintainable thing.

The primary case where it is beneficial to split based on checking IsCompleted{Successfully} on a ValueTask is when you need to do additional work only in the case where it's still running asynchronously, e.g. if you can delay registering for cancellation until you know the operation didn't complete synchronously.  That doesn't appear to be the case here.

Reply:
thanks for detailed explanation @stephentoub. I turn it into async function for now to simplify the code.